--- compat-wireless-2014-01-23.1/drivers/net/wireless/ath/ath9k/beacon.c.orig	2014-03-06 16:04:12.744389433 +0100
+++ compat-wireless-2014-01-23.1/drivers/net/wireless/ath/ath9k/beacon.c	2014-03-06 16:07:07.075205012 +0100
@@ -332,7 +332,7 @@ void ath9k_beacon_tasklet(unsigned long
 	 * a problem and should not occur.  If we miss too
 	 * many consecutive beacons reset the device.
 	 */
-	if (ath9k_hw_numtxpending(ah, sc->beacon.beaconq) != 0) {
+	if (ath9k_hw_numtxpending(ah, sc->beacon.beaconq) >= ATH_BCBUF) {
 		sc->beacon.bmisscnt++;
 
 		ath9k_hw_check_nav(ah);
