--- a/net/bridge/netfilter/nft_reject_bridge.c
+++ b/net/bridge/netfilter/nft_reject_bridge.c
@@ -31,6 +31,13 @@ static void nft_reject_br_push_etherhdr(
 	ether_addr_copy(eth->h_dest, eth_hdr(oldskb)->h_source);
 	eth->h_proto = eth_hdr(oldskb)->h_proto;
 	skb_pull(nskb, ETH_HLEN);
+
+	if (skb_vlan_tag_present(oldskb)) {
+		u16 vid = skb_vlan_tag_get(oldskb);
+		__vlan_hwaccel_put_tag(nskb, oldskb->vlan_proto, vid);
+		//skb = vlan_insert_tag_set_proto(nskb, oldskb->vlan_proto,
+		//				skb_vlan_tag_get(oldskb));
+	}
 }
 
 static int nft_bridge_iphdr_validate(struct sk_buff *skb)
